opm/autodiff/StandardWells_impl.hpp

```diff
@@ -1248,4 +1248,304 @@ namespace Opm
+    bool
+    StandardWells::
+    checkRateEconLimits(const WellEconProductionLimits& econ_production_limits,
+                        const WellState& well_state,
+                        const int well_number) const
+    {
+        const Opm::PhaseUsage& pu = fluid_->phaseUsage();
+        const int np = well_state.numPhases();
+
+        if (econ_production_limits.onMinOilRate()) {
+            assert((*active_)[Oil]);
+            const double oil_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Oil ] ];
+            const double min_oil_rate = econ_production_limits.minOilRate();
+            if (std::abs(oil_rate) < min_oil_rate) {
+                return true;
+            }
+        }
+
+        if (econ_production_limits.onMinGasRate() ) {
+            assert((*active_)[Gas]);
+            const double gas_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Gas ] ];
+            const double min_gas_rate = econ_production_limits.minGasRate();
+            if (std::abs(gas_rate) < min_gas_rate) {
+                return true;
+            }
+        }
+
+        if (econ_production_limits.onMinLiquidRate() ) {
+            assert((*active_)[Oil]);
+            assert((*active_)[Water]);
+            const double oil_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Oil ] ];
+            const double water_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Water ] ];
+            const double liquid_rate = oil_rate + water_rate;
+            const double min_liquid_rate = econ_production_limits.minLiquidRate();
+            if (std::abs(liquid_rate) < min_liquid_rate) {
+                return true;
+            }
+        }
+
+        if (econ_production_limits.onMinReservoirFluidRate()) {
+            OpmLog::warning("Minimum reservoir fluid production rate limit is not supported yet");
+        }
+
+        return false;
+    }
+
+
+
+
+
+    template <class WellState>
+    bool
+    StandardWells::
+    checkRatioEconLimits(const WellEconProductionLimits& econ_production_limits,
+                         const WellState& well_state,
+                         const typename WellMapType::const_iterator& i_well,
+                         int& worst_offending_connection,
+                         bool& last_connection) const
+    {
+        // TODO: not sure how to define the worst-offending connection when more than one
+        //       ratio related limit is violated.
+        //       The defintion used here is that we define the violation extent based on the
+        //       ratio between the value and the corresopoding limit.
+        //       For each violated limit, we decide the worst-offending connection separately.
+        //       Among the worst-offending connections, we use the one has the biggest violation
+        //       extent.
+
+        bool any_limit_violated = false;
+        double violation_extent = 0.0;
+        worst_offending_connection = -1;
```

Comment:
Instead of using -1, you should name the constant in an enum (and use it also in the function below).